docs/node2node.md

```diff
@@ -0,0 +1,105 @@
+# node to node messages
+
+
+[summary placeholder]
+
+(todo: fully describe pop and shift mechanism)
+(todo: calcuate optimal data sizes)
+
+
+## streamed requests
+
+direct node to node request/response messages sent on a bidi quic stream.
+
+the stream starts by sending 2 bytes identifying the stream kind of which there are the following:
+
+* network data lookup
+* network data upsert
+* path build
+* path update
+* path stream
+
+### network data lookup stream
+
+used for fetching of:
+
+* pubkey list with unlock times if they are applicable.
+* pubkey list delta given last hash of full list.
+* router contact by snode identity pubkey.
+* router contacts by operator wallet address.
+* look up wallet addresses staked into snode.
+* look up snodes staked by a wallet address.
+
+
+### network data upsert stream
+
+used to upsert (insert/update) records for:
+
+* gossiped router contacts.
+* encrypted introsets.
+* signed service descriptors. (aka the "exit market" metadata).
+
+
+### path builds
+
+#### path build request
+
+we have 8 path build records in a path build, we pop and shift the records and decrypt the record meant for us.
+
+the plainext record contains:
+
+* next hop's pubkey
+* path lifetime
+* signing pubkey used by the path creator to sign updates
+* encryption pubkey to generate shared secret at the path
+* additional extra fields used in the future.
+
+
+(todo: i want to permit multiple kinds of pubkeys for forwards compat)
+
+we could also add an additional list of nonces on the path build that lets us open multiple path streams as we build our path.
+given we provide N nonces in the build we are opening up N path streams as we make our path build. see the path streams section below.
+
+(todo: shrink tx/rx path id to be 8 bytes each?)
+
+#### path build result
+
+(todo: finish this)
+
+### path streams
+
+onion routed persisting stream connection to the pivot router for sending anonymised requests destined for the pivot router.
+
+streams are created for an existing path by opening a "path stream" stream providing the path's tx/rx id and 24 bytes nonce.
+on accepting the stream creation the node will open a "path stream" to their upstream router but with a nonce that is deterministically mutated `new_nonce = nonce ^ hash(nonce)`, this mutation reduces stream corrilation across hops.
```

Comment:
this operation needs to be commutative as it is done in both directions. when we get a nonce as the path builder from the pivot router we need to be able to do each operation in reverse order.

Reply:
I'm not understanding.

If I compute:

    X = N xor hash(N)

then how do I get N from X?

Reply:
that was my point, you can't.

Reply:
So then I'm back to my original comment: `X = N xor hash(N)` offers nothing over just `X = hash(N)`

Reply:
they both suffer from the same problem. it's a one way operation. when the sender of the original message is not the path creator we cannot obtain the original value of X because it's gone through a hash function. what we do right now with our existing scheme is `N = N ^ hash(path_shared_secret_symettric_key)`, note that the operation applied is `^`, and the `hash()` part of it is from the frame of reference of the next hop they dont really care that a hash even happens. this is a tangent, i'm redoing this entire pr as i misunderstood the level of detail required.